lib/pipeline_controller.py

```diff
@@ -73,9 +126,99 @@ def process_all_sites_data(self, end_date, batch_num):
+        # particular date (e.g. when only one of several orbits is broken, or when an
+        # orbit goes down in the middle of the day).
+        known_data_dict = dict()
+        if known_data:
+            known_data_dict = {
+                (row[0], row[1], row[2]): (row[3], row[4], row[5], row[6])
+                for row in known_data
+            }
+        self._recover_data(recoverable_site_dates, known_data_dict)
+        self.redshift_client.close_connection()
+
+    def _recover_data(self, site_dates, known_data_dict):
+        """
+        Individually query the ShopperTrak API for each site/date pair with any
+        unhealthy data. Then check to see if the returned data is actually "recovered"
+        data, as it may have never been unhealthy to begin with. If so, send to Kinesis.
+        """
+        for row in site_dates:
+            site_xml_root = self.shoppertrak_api_client.query(
+                SINGLE_SITE_ENDPOINT + row[0], row[1]
+            )
+            if site_xml_root:
+                site_results = self.shoppertrak_api_client.parse_response(
+                    site_xml_root, row[1], is_recovery_mode=True
+                )
+                self._process_recovered_data(site_results, known_data_dict)
+
+    def _process_recovered_data(self, recovered_data, known_data_dict):
+        """
+        Check that ShopperTrak "recovered" data was actually unhealthy to begin with
+        and, if so, encode, send to Kinesis, and mark old Redshift rows as stale
+        """
+        results = []
+        stale_ids = []
+        for fresh_row in recovered_data:
+            key = (
+                fresh_row["shoppertrak_site_id"],
+                fresh_row["orbit"],
+                datetime.strptime(fresh_row["increment_start"], "%Y-%m-%d %H:%M:%S"),
+            )
+            if key not in known_data_dict:
+                results.append(fresh_row)
+            else:
+                known_row = known_data_dict[key]
+                if not known_row[1]:  # previously unhealthy data
+                    results.append(fresh_row)
+                    stale_ids.append(str(known_row[0]))
+                elif (  # previously healthy data that doesn't match the new API data
+                    fresh_row["enters"] != known_row[2]
+                    or fresh_row["exits"] != known_row[3]
+                ):
+                    self.logger.warning(
+                        f"Different healthy data found in API and Redshift: {key} "
+                        f"mapped to {fresh_row} in the API and {known_row} in Redshift"
+                    )
+
+        # Mark old rows for successfully recovered data as stale
+        update_query = build_redshift_update_query(
+            self.redshift_visits_table, ",".join(stale_ids)
+        )
+        self.redshift_client.execute_transaction([(update_query, None)])
+
+        if results:
+            encoded_records = self.avro_encoder.encode_batch(results)
```

Comment:
Will Avro validate these records? They look like they're missing fields from the "full" ones.

Reply:
oh, hmm, maybe I misread this but I thought the `fresh_row` objects were only getting a subset of the fields.